components/script/dom/globalscope.rs

```diff
@@ -207,6 +357,369 @@ impl GlobalScope {
+                let mut entry = potential_gc.entry(id.clone()).or_insert((1, 1));
+                // Each time entry.0 reaches entry.1, we send a message to check if we can GC.
+                // To avoid sending too many messages, entry.1 doubles each time it is reached,
+                // while entry.0 increments at each GC checkpoint.
+                if entry.0 == entry.1 {
+                    entry.1 = entry.1.checked_mul(2).unwrap_or(u32::max_value());
+                    port.send_potential_gc_msg();
+                }
+                entry.0 = entry.0.checked_add(1).unwrap_or(0);
+            }
+        }
+        for id in to_be_removed.drain() {
+            self.remove_message_port(id);
+        }
+    }
+
+    /// Handle the case were our entangled port is ready for GC,
+    /// and wants to know if we are still used.
+    pub fn handle_check_gc_messageport(&self, port_id: MessagePortId) {
+        let alive_js = match self.message_port_tracker.borrow().get(&port_id) {
+            Some(weak) => weak.root().is_some(),
+            None => false,
+        };
+        if !alive_js {
+            if let Some(port) = self.message_ports.borrow().get(&port_id) {
+                port.comfirm_gc();
+            }
+        }
+    }
+
+    /// Handle the case were our entangled port has confirmed to us we can GC.
+    pub fn handle_comfirmed_gc_messageport(&self, port_id: MessagePortId) {
+        let alive_js = match self.message_port_tracker.borrow().get(&port_id) {
+            Some(weak) => weak.root().is_some(),
+            None => false,
+        };
+        // Check again, in case a message was in transit and received after we last checked.
+        let still_possibly_unreachable =
+            if let Some(port) = self.message_ports.borrow().get(&port_id) {
+                port.possibly_unreachable()
+            } else {
+                false
+            };
+        if !alive_js && still_possibly_unreachable {
+            self.remove_message_port(port_id.clone());
+            // Let the constellation know to drop this port and the one it is entangled with,
+            // and to forward this message to the script-process where the entangled is found.
+            let _ = self
+                .script_to_constellation_chan()
+                .send(ScriptMsg::RemoveMessagePort(port_id.clone()));
+        }
+    }
+
+    /// Get a DOM object corresponding to this MessagePortId.
+    fn get_dom_message_port(&self, port_id: &MessagePortId) -> DomRoot<MessagePort> {
+        let listener = self.get_message_port_onmessage(port_id);
+        let mut dom_ports = self.message_port_tracker.borrow_mut();
+        let weak = dom_ports.get(port_id).expect("MessagePort to be tracked");
+        let dom_port = match weak.root() {
+            Some(dom_port) => dom_port,
+            None => {
+                // If the current one has been GC'ed, we create a new one.
```

Comment:
On this one, from my experience with debugging this, it appears that ports are GC'ed even when a message handler is attached to them. 

The default "use" of ports in the wpt tests seems to be getting a port as a transfer argument inside a handler for `MessageEvent`, and then simply attaching an `onmessage` handler to the port and letting it "drop". 

In such a case, unless Servo keeps a `Dom`, the port will be GC'ed and tests start failing because the `onmessage` handler expects to be called even when the JS port has been GC'ed. 

So if we panic here, we'll be panicking quite often in the "normal course of business" for message-ports. 

Also note that the `onmessage` handlers are stored separately, and restored on the new objects created here(although not generic event-handlers). 

And you're right that the new object is only created here if the previous one is unavailable from JS, hence no equality check in JS would be possible. 

So what is happening here is that we create a new port, and then set it as the `target` of the `MessageEvent` we are about to dispatch, and we only create a new one is the previous one has been GC'ed. 

The other option I epxerimented with is basically keeping `Dom`, and then the problem I had is that I couldn't actually know when the corresponding JS object was "ready for GC on the JS side of things". I tried doing some stuff in the tracing, but that is called "too early"...

Reply:
The spec also mentions this as a special case of garbage collection, which I think implies the Web engine needs to do something in addition to the JS engine, since the default GC behavior will not work out of the box: https://html.spec.whatwg.org/multipage/web-messaging.html#ports-and-garbage-collection

Reply:
Note that the event-target of the dom object is traced at https://github.com/servo/servo/pull/23637/files#diff-7a9812c344e5a2ce6252e367e0297530R52

However, I just realize I might be ommitting setting the handler on the event-target, since I only store it at https://github.com/servo/servo/pull/23637/files#diff-59d233642d0ce6d687484bdd009e1017R457

So it might be that I skipped setting the `onmessage` handler on the current event-target, which is causing the early GC due to Spidermonkey not knowning about the handler? It's something I will have to try out because I'm not sure in what state the code was when I first noted the "early GC" behavior noted above.